File: lexer.ts
import { Token, lookupIdent } from "./token.ts";

function isLetter(ch: string): boolean {
  return ("a" <= ch && ch <= "z") || ("A" <= ch && ch <= "Z") || ch === "_";
}

function isDigit(ch: string): boolean {
  return "0" <= ch && ch <= "9";
}

export class Lexer {
  #input: string;
  #position = 0;
  #readPosition = 0;
  #ch = "";

  public constructor(input: string) {
    this.#input = input;
    this.#readChar();
  }

  public nextToken(): Token {
    let token: Token;

    this.#skipWhitespace();

    switch (this.#ch) {
      case "=":
        if (this.#peekChar() === "=") {
          const ch = this.#ch;
          this.#readChar();
          const literal = ch + this.#ch;
          token = new Token("EQ", literal);
        } else {
          token = new Token("ASSIGN", this.#ch);
        }
        break;
      case "+":
        token = new Token("PLUS", this.#ch);
        break;
      case "-":
        token = new Token("MINUS", this.#ch);
        break;
      case "!":
        if (this.#peekChar() === "=") {
          const ch = this.#ch;
          this.#readChar();
          const literal = ch + this.#ch;
          token = new Token("NOT_EQ", literal);
        } else {
          token = new Token("BANG", this.#ch);
        }
        break;
      case "/":
        token = new Token("SLASH", this.#ch);
        break;
      case "*":
        token = new Token("ASTERISK", this.#ch);
        break;
      case "<":
        token = new Token("LT", this.#ch);
        break;
      case ">":
        token = new Token("GT", this.#ch);
        break;
      case ";":
        token = new Token("SEMICOLON", this.#ch);
        break;
      case "(":
        token = new Token("LPAREN", this.#ch);
        break;
      case ")":
        token = new Token("RPAREN", this.#ch);
        break;
      case ",":
        token = new Token("COMMA", this.#ch);
        break;
      case "{":
        token = new Token("LBRACE", this.#ch);
        break;
      case "}":
        token = new Token("RBRACE", this.#ch);
        break;
      case '"':
        token = new Token("STRING", this.#readString());
        break;
      case "\0":
        token = new Token("EOF", "");
        break;
      default:
        if (isLetter(this.#ch)) {
          const literal = this.#readIdentifier();
          const type = lookupIdent(literal);
          return new Token(type, literal);
        } else if (isDigit(this.#ch)) {
          const literal = this.#readNumber();
          return new Token("INT", literal);
        } else {
          token = new Token("ILLEGAL", this.#ch);
        }
    }

    this.#readChar();
    return token;
  }

  #peekChar(): string {
    if (this.#readPosition >= this.#input.length) {
      return "\0";
    } else {
      return this.#input[this.#readPosition]!;
    }
  }

  #readChar(): void {
    if (this.#readPosition >= this.#input.length) {
      this.#ch = "\0";
    } else {
      this.#ch = this.#input[this.#readPosition]!;
    }

    this.#position = this.#readPosition;
    this.#readPosition += 1;
  }

  #readIdentifier(): string {
    const position = this.#position;
    while (isLetter(this.#ch)) {
      this.#readChar();
    }
    return this.#input.slice(position, this.#position);
  }

  #readNumber(): string {
    const position = this.#position;
    while (isDigit(this.#ch)) {
      this.#readChar();
    }
    return this.#input.slice(position, this.#position);
  }

  #readString(): string {
    const position = this.#position + 1;
    while (true) {
      this.#readChar();
      if (this.#ch === '"' || this.#ch === "\0") {
        break;
      }
    }
    return this.#input.slice(position, this.#position);
  }

  #skipWhitespace(): void {
    while (
      this.#ch === " " ||
      this.#ch === "\t" ||
      this.#ch === "\n" ||
      this.#ch === "\r"
    ) {
      this.#readChar();
    }
  }
}

File: rust/src/lexer/lexer.rs
use anyhow::Result;

#[derive(Debug, PartialEq)]
pub enum Token {
    Illegal,
    Eof,

    // Identifiers + literals
    Ident(String),
    Int(String),

    // Operators
    Assign,
    Plus,
    Minus,
    Bang,
    Asterisk,
    Slash,
    Lt,
    Gt,
    Eq,
    NotEq,

    // Delimiters
    Comma,
    Semicolon,

    LParen,
    RParen,
    LBrace,
    RBrace,

    // Keywords
    Function,
    Let,
    If,
    Else,
    Return,
    True,
    False,
}

pub struct Lexer {
    input: Vec<u8>,
    position: usize,
    read_position: usize,
    ch: u8,
}

impl Lexer {
    pub fn new(input: String) -> Lexer {
        let mut lexer = Lexer {
            input: input.into_bytes(),
            position: 0,
            read_position: 0,
            ch: 0,
        };
        lexer.read_char();
        return lexer;
    }

    pub fn next_token(&mut self) -> Result<Token> {
        self.skip_whitespace();

        let token = match self.ch {
            b'=' => {
                if self.peek_char() == b'=' {
                    self.read_char();
                    Token::Eq
                } else {
                    Token::Assign
                }
            }
            b'+' => Token::Plus,
            b'-' => Token::Minus,
            b'!' => {
                if self.peek_char() == b'=' {
                    self.read_char();
                    Token::NotEq
                } else {
                    Token::Bang
                }
            }
            b'*' => Token::Asterisk,
            b'/' => Token::Slash,
            b'<' => Token::Lt,
            b'>' => Token::Gt,
            b';' => Token::Semicolon,
            b'(' => Token::LParen,
            b')' => Token::RParen,
            b',' => Token::Comma,
            b'{' => Token::LBrace,
            b'}' => Token::RBrace,
            b'a'..=b'z' | b'A'..=b'Z' | b'_' => {
                let identifier = self.read_identifier();
                return Ok(match identifier.as_str() {
                    "fn" => Token::Function,
                    "let" => Token::Let,
                    "if" => Token::If,
                    "else" => Token::Else,
                    "return" => Token::Return,
                    "true" => Token::True,
                    "false" => Token::False,
                    _ => Token::Ident(identifier),
                });
            }
            b'0'..=b'9' => {
                let number = self.read_number();
                return Ok(Token::Int(number));
            }
            0 => Token::Eof,
            _ => Token::Illegal,
        };

        self.read_char();
        return Ok(token);
    }

    fn read_char(&mut self) {
        if self.read_position >= self.input.len() {
            self.ch = 0;
        } else {
            self.ch = self.input[self.read_position];
        }

        self.position = self.read_position;
        self.read_position += 1;
    }

    fn peek_char(&mut self) -> u8 {
        if self.read_position >= self.input.len() {
            return 0;
        } else {
            return self.input[self.read_position];
        }
    }

    fn skip_whitespace(&mut self) {
        while self.ch.is_ascii_whitespace() {
            self.read_char();
        }
    }

    fn read_identifier(&mut self) -> String {
        let position = self.position;
        while self.ch.is_ascii_alphabetic() || self.ch == b'_' {
            self.read_char();
        }
        return String::from_utf8_lossy(&self.input[position..self.position]).to_string();
    }

    fn read_number(&mut self) -> String {
        let position = self.position;
        while self.ch.is_ascii_digit() {
            self.read_char();
        }
        return String::from_utf8_lossy(&self.input[position..self.position]).to_string();
    }
}

#[cfg(test)]
mod tests {
    use super::{Lexer, Token};
    use anyhow::Result;

    #[test]
    fn test_next_token() -> Result<()> {
        let input = "let five = 5;
let ten = 10;

let add = fn(x, y) {
    x + y;
};

let result = add(five, ten);
!-/*5;
5 < 10 > 5;

if (5 < 10) {
    return true;
} else {
    return false;
}

10 == 10;
10 != 9;
";

        let tests = vec![
            Token::Let,
            Token::Ident(String::from("five")),
            Token::Assign,
            Token::Int(String::from("5")),
            Token::Semicolon,
            Token::Let,
            Token::Ident(String::from("ten")),
            Token::Assign,
            Token::Int(String::from("10")),
            Token::Semicolon,
            Token::Let,
            Token::Ident(String::from("add")),
            Token::Assign,
            Token::Function,
            Token::LParen,
            Token::Ident(String::from("x")),
            Token::Comma,
            Token::Ident(String::from("y")),
            Token::RParen,
            Token::LBrace,
            Token::Ident(String::from("x")),
            Token::Plus,
            Token::Ident(String::from("y")),
            Token::Semicolon,
            Token::RBrace,
            Token::Semicolon,
            Token::Let,
            Token::Ident(String::from("result")),
            Token::Assign,
            Token::Ident(String::from("add")),
            Token::LParen,
            Token::Ident(String::from("five")),
            Token::Comma,
            Token::Ident(String::from("ten")),
            Token::RParen,
            Token::Semicolon,
            Token::Bang,
            Token::Minus,
            Token::Slash,
            Token::Asterisk,
            Token::Int(String::from("5")),
            Token::Semicolon,
            Token::Int(String::from("5")),
            Token::Lt,
            Token::Int(String::from("10")),
            Token::Gt,
            Token::Int(String::from("5")),
            Token::Semicolon,
            Token::If,
            Token::LParen,
            Token::Int(String::from("5")),
            Token::Lt,
            Token::Int(String::from("10")),
            Token::RParen,
            Token::LBrace,
            Token::Return,
            Token::True,
            Token::Semicolon,
            Token::RBrace,
            Token::Else,
            Token::LBrace,
            Token::Return,
            Token::False,
            Token::Semicolon,
            Token::RBrace,
            Token::Int(String::from("10")),
            Token::Eq,
            Token::Int(String::from("10")),
            Token::Semicolon,
            Token::Int(String::from("10")),
            Token::NotEq,
            Token::Int(String::from("9")),
            Token::Semicolon,
            Token::Eof,
        ];

        let mut lexer = Lexer::new(input.to_string());

        for expected_token in tests {
            let token = lexer.next_token()?;
            assert_eq!(token, expected_token);
        }

        return Ok(());
    }
}
